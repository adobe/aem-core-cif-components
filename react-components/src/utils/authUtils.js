/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import { ApolloLink } from '@apollo/client';
import { checkCookie, cookieValue } from './cookieUtils';

const authLink = new ApolloLink((operation, forward) => {
    let token = checkCookie('cif.userToken') ? cookieValue('cif.userToken') : '';
    if (token.length > 0) {
        operation.setContext(({ headers }) => ({
            headers: {
                authorization: `Bearer ${token && token.length > 0 ? token : ''}`,
                ...headers
            }
        }));
    }

    return forward(operation);
});

export { authLink as graphqlAuthLink };
