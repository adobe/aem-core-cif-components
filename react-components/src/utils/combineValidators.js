/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

/**
 * The function purpose is to group validation callbacks into a chain within single callback function.
 *
 * The single callback function should be passed as `validation` prop to an input component. It's being
 * called by the React Controller on a form submit.
 * The `callbacks` param is the array contains validators. Each validator should be either function or array. In case of a
 * validator is a function it is called the same way as React Controller does, with a field `value` and a form `values` params. If
 * the validator is an array, this means that it is extended validator which requires additional param for configuration.
 * The first index of that array should be a extended validator function, the second index is extended param, which is needed for
 * that function. So the function is called with three params on the form submit, with a field `value`, a form `values`
 * and an `extended param`.
 *
 * Example usage of the function:
 *
 * <input validate={combine([
 * foo
 * ])} .../>
 *
 * foo - validation function
 * It will be called on form submitting within chain as:
 * `foo(value, values)`
 *
 * Example usage of the function with extended callback:
 *
 * <input validate={combine([
 * [foo, bar]
 * ])} .../>
 *
 * foo - extended validator
 * bar - additional param
 * It will be called on form submitting within chain as:
 * `foo(value, values, bar)`
 *
 * Each callback being called one after another, according to their index in callbacks array, if one item returned the error
 * message, validation is failed, and rest validator are not supposed to be called.
 *
 * @param {Array} callbacks
 * @return {function(value, values): *}
 */
export default callbacks => {
    if (callbacks == null || !Array.isArray(callbacks)) {
        throw new Error('Expected `callbacks` to be array.');
    }

    return (value, values) => {
        let result = null;

        for (let i = 0; i < callbacks.length; i++) {
            const callback = callbacks[i];

            if (callback == null || (!Array.isArray(callback) && typeof callback !== 'function')) {
                throw new Error('Expected `callbacks[' + i + ']` to be array or function.');
            }

            if (Array.isArray(callback)) {
                const [extendedCallback, extendedParam] = callback;

                if (typeof extendedCallback !== 'function') {
                    throw new Error('Expected `callbacks[' + i + '][0]` to be function.');
                }

                result = extendedCallback(value, values, extendedParam);
            } else {
                result = callback(value, values);
            }

            if (result) {
                break;
            }
        }

        return result;
    };
};
