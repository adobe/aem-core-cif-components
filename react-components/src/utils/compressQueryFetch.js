/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import { stripIgnoredCharacters } from 'graphql';

/**
 * This custom fetch interceptor minimizes GraphQL queries, by stripping unnecessary characters from the query.
 * This is a workaround, since Apollo always restores the formatting of a query, independent of the formatting in
 * .graphql files. See https://github.com/apollographql/apollo-link/issues/1079.
 *
 * This is required, since there are certain limits for Magento Cloud, see
 * https://docs.fastly.com/en/guides/resource-limits#request-and-response-limits.
 */
export default function(url, options) {
    try {
        if (options.method === 'GET') {
            // Split url by ? to get query string
            let parts = url.split('?');

            // Only proceed if it is a valid URL
            if (parts.length === 2) {
                // Parse query from search params string
                let params = new URLSearchParams(parts[1]);
                let query = params.get('query');
                if (query) {
                    query = stripIgnoredCharacters(query);
                    params.set('query', query);
                    parts[1] = params.toString();
                    url = parts.join('?');
                }
            }
        }
        if (options.method === 'POST') {
            // Parse query from body in request
            let bodyJson = JSON.parse(options.body);
            if (bodyJson.query) {
                bodyJson.query = stripIgnoredCharacters(bodyJson.query);
                options.body = JSON.stringify(bodyJson);
            }
        }
    } catch (err) {
        console.warn('Could not minimize GraphQL query', err);
    }
    return fetch(url, options);
}
