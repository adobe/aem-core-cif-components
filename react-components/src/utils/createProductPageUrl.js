/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

let storeRootUrl = null;

export const createProductPageUrl = sku => {
    const url = new URL(window.location);

    // Provided by StoreConfigExporter
    if (storeRootUrl) {
        return storeRootUrl;
    }

    let storeConfigEl = document.querySelector('meta[name="store-config"]');
    let pathname;

    if (storeConfigEl) {
        pathname = JSON.parse(storeConfigEl.content).storeRootUrl;
    } else {
        // TODO: deprecated - the store configuration on the <body> has been deprecated and will be removed
        pathname = document.body.dataset.storeRootUrl;
    }

    if (!pathname) {
        return null;
    }

    const extension = '.html';
    const path = pathname.substr(0, pathname.lastIndexOf('.'));

    url.pathname = `${path}.cifproductredirect${extension}/${sku}`;

    return (storeRootUrl = url.toString());
};
