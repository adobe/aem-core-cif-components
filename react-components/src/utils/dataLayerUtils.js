/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
const dataLayerEnabled = document.body.hasAttribute('data-cmp-data-layer-enabled');
const dataLayer = dataLayerEnabled ? (window.adobeDataLayer = window.adobeDataLayer || []) : undefined;

// https://github.com/adobe/adobe-client-data-layer/wiki#push
export const pushData = data => {
    if (dataLayerEnabled) {
        dataLayer.push(data);
    }
};

// https://github.com/adobe/adobe-client-data-layer/wiki#push
export const pushEvent = (eventName, eventData) => {
    if (dataLayerEnabled) {
        dataLayer.push({
            event: eventName,
            eventInfo: eventData
        });
    }
};

// https://github.com/adobe/adobe-client-data-layer/wiki#getstate
export const getState = reference => {
    if (dataLayerEnabled) {
        return dataLayer.getState(reference);
    }
    return null;
};

// https://github.com/adobe/adobe-client-data-layer/wiki#addeventlistener
export const addEventListener = (eventName, handler) => {
    if (dataLayerEnabled) {
        dataLayer.addEventListener(eventName, handler);
    }
};

// https://github.com/adobe/adobe-client-data-layer/wiki#removeeventlistener
export const removeEventListener = (eventName, handler) => {
    if (dataLayerEnabled) {
        dataLayer.removeEventListener(eventName, handler);
    }
};
