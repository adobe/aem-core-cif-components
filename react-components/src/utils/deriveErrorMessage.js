/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

/**
 * Maps an error to a string message
 *
 * @param {Error} error the error to map
 * @return {String} error message
 */
const toString = error => {
    const { graphQLErrors, message } = error;

    return graphQLErrors && graphQLErrors.length ? graphQLErrors.map(({ message }) => message).join(', ') : message;
};

/**
 * A function to derive an error string from an array of errors.
 */
export const deriveErrorMessage = errors => {
    const errorCollection = [];
    for (const error of errors) {
        if (error) {
            errorCollection.push(toString(error));
        }
    }

    return errorCollection.join(', ');
};
