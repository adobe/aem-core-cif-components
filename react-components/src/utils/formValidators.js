/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

const SUCCESS = undefined;

export const hasLengthAtLeast = (value, values, minimumLength) => {
    if (!value || value.length < minimumLength) {
        return `Must contain at least ${minimumLength} character(s).`;
    }

    return SUCCESS;
};

export const hasLengthAtMost = (value, values, maximumLength) => {
    if (value && value.length > maximumLength) {
        return `Must not exceed ${maximumLength} character(s).`;
    }

    return SUCCESS;
};

export const hasLengthExactly = (value, values, length) => {
    if (value && value.length !== length) {
        return `Must contain exactly ${length} character(s).`;
    }

    return SUCCESS;
};

export const isRequired = value => {
    return (value || '').trim() ? SUCCESS : 'The field is required.';
};

export const validateEmail = value => {
    const regex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    return regex.test(value) ? SUCCESS : 'Please enter a valid email address (Ex: johndoe@domain.com).';
};

export const validateRegionCode = (value, values, countries) => {
    const country = countries.find(({ id }) => id === 'US');

    if (!country) {
        return 'Country "US" is not an available country.';
    }
    const { available_regions: regions } = country;

    if (!(Array.isArray(regions) && regions.length)) {
        return 'Country "US" does not contain any available regions.';
    }
    const region = regions.find(({ code }) => code === value);
    if (!region) {
        return `State "${value}" is not an valid state abbreviation.`;
    }

    return SUCCESS;
};

export const validatePassword = value => {
    const count = {
        lower: 0,
        upper: 0,
        digit: 0,
        special: 0
    };

    for (const char of value) {
        if (/[a-z]/.test(char)) count.lower++;
        else if (/[A-Z]/.test(char)) count.upper++;
        else if (/\d/.test(char)) count.digit++;
        else if (/\S/.test(char)) count.special++;
    }

    if (Object.values(count).filter(Boolean).length < 3) {
        return 'A password must contain at least 3 of the following: lowercase, uppercase, digits, special characters.';
    }

    return SUCCESS;
};

export const validateConfirmPassword = (value, values, passwordKey = 'password') => {
    return value === values[passwordKey] ? SUCCESS : 'Passwords must match.';
};

export const isNotEqualToField = (value, values, otherField) => {
    return value !== values[otherField] ? SUCCESS : `${otherField} must be different`;
};
