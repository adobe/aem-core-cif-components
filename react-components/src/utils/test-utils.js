/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/
import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { I18nextProvider } from 'react-i18next';

import i18n from '../../__mocks__/i18nForTests';
import ConfigContextProvider from '../context/ConfigContext';
import UserContextProvider from '../context/UserContext';

import mutationChangePassword from './mocks/mutationChangePassword';
import mutationCreateCustomer from './mocks/mutationCreateCustomer';
import mutationCreateDuplicateCustomer from './mocks/mutationCreateCustomerDuplicate';
import mutationDeleteCustomerAddress from './mocks/mutationDeleteCustomerAddress';
import mutationGenerateToken from './mocks/mutationGenerateToken';
import mutationGenerateTokenWrongPassword from './mocks/mutationGenerateTokenWrongPassword';
import mutationMergeCarts from './mocks/mutationMergeCarts';
import mutationPlaceOrder from './mocks/mutationPlaceOrder';
import mutationRevokeToken from './mocks/mutationRevokeToken';
import mutationShippingAddress from './mocks/mutationShippingAddress';
import queryCart from './mocks/queryCart';
import queryCountries from './mocks/queryCountries';
import queryCustomerCart from './mocks/queryCustomerCart';
import queryCustomerDetails from './mocks/queryCustomerDetails';
import queryCustomerInformation from './mocks/queryCustomerInformation';
import queryEmptyCart from './mocks/queryEmptyCart';
import queryNewCart from './mocks/queryNewCart';

const defaultMocks = [
    mutationChangePassword,
    mutationCreateCustomer,
    mutationCreateDuplicateCustomer,
    mutationDeleteCustomerAddress,
    mutationGenerateToken,
    mutationGenerateTokenWrongPassword,
    mutationMergeCarts,
    mutationPlaceOrder,
    mutationRevokeToken,
    mutationShippingAddress,
    queryCart,
    queryCountries,
    queryCustomerCart,
    queryCustomerDetails,
    queryCustomerInformation,
    queryEmptyCart,
    queryNewCart
];

const defaultConfig = {
    storeView: 'default',
    graphqlEndpoint: 'none'
};

// eslint-disable-next-line react/display-name
const allProviders = (config, userContext, mocks) => ({ children }) => {
    return (
        <MockedProvider mocks={mocks} addTypename={false}>
            <ConfigContextProvider config={config || defaultConfig}>
                <UserContextProvider initialState={userContext}>
                    <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
                </UserContextProvider>
            </ConfigContextProvider>
        </MockedProvider>
    );
};

/* Wrap all the React components tested with the library in a mocked Apollo provider */
const customRender = (ui, options = {}) => {
    const { config, userContext, mocks = defaultMocks, ...renderOptions } = options;
    return render(ui, { wrapper: allProviders(config, userContext, mocks), ...renderOptions });
};

export * from '@testing-library/react';
export { customRender as render };
