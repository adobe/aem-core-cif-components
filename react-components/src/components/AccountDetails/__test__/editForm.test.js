/*******************************************************************************
 *
 *    Copyright 2020 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React from 'react';
import { render } from '../../../utils/test-utils';
import EditForm from '../editForm';

jest.mock('informed', () => {
    const informed = jest.requireActual('informed');

    return {
        ...informed,
        useFormApi: () => {
            return {
                setValue: () => {}
            };
        }
    };
});

describe('<EditForm>', () => {
    afterAll(() => {
        jest.clearAllMocks();
    });

    it('renders a form with the "Change password" button', () => {
        const { asFragment } = render(<EditForm />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders a form without the "Change password" button', () => {
        const { asFragment } = render(<EditForm shouldShowNewPassword={true} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
