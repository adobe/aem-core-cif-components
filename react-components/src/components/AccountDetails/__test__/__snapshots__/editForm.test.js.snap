// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditForm> renders a form with the "Change password" button 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="firstname"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          First Name
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              class="input"
              name="firstname"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="lastname"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          Last Name
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              class="input"
              name="lastname"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="email"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          Email address
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              class="input"
              name="email"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="password"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          Password
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              autocomplete="current-password"
              class="input"
              name="password"
              type="password"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          >
            <button
              type="button"
            >
              <span
                class="content"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17.94 17.94A10.07 10.07 0 0 1 12 20c-7 0-11-8-11-8a18.45 18.45 0 0 1 5.06-5.94M9.9 4.24A9.12 9.12 0 0 1 12 4c7 0 11 8 11 8a18.5 18.5 0 0 1-2.16 3.19m-6.72-1.07a3 3 0 1 1-4.24-4.24"
                  />
                  <line
                    x1="1"
                    x2="23"
                    y1="1"
                    y2="23"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="changePasswordButtonContainer"
    >
      <button
        class="changePasswordButton"
        classes="[object Object]"
        type="button"
      >
        <span>
          Change Password
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<EditForm> renders a form without the "Change password" button 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="firstname"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          First Name
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              class="input"
              name="firstname"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="lastname"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          Last Name
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              class="input"
              name="lastname"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="email"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          Email address
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              class="input"
              name="email"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          />
        </span>
        <p
          class="root"
        />
      </div>
    </div>
    <div
      class="password"
    >
      <div
        class="root"
      >
        <span
          class="label"
        >
          Current Password
        </span>
        <span
          class="root"
        >
          <span
            class="input"
          >
            <input
              autocomplete="current-password"
              class="input"
              name="password"
              type="password"
              value=""
            />
          </span>
          <span
            class="before"
          />
          <span
            class="after"
          >
            <button
              type="button"
            >
              <span
                class="content"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17.94 17.94A10.07 10.07 0 0 1 12 20c-7 0-11-8-11-8a18.45 18.45 0 0 1 5.06-5.94M9.9 4.24A9.12 9.12 0 0 1 12 4c7 0 11 8 11 8a18.5 18.5 0 0 1-2.16 3.19m-6.72-1.07a3 3 0 1 1-4.24-4.24"
                  />
                  <line
                    x1="1"
                    x2="23"
                    y1="1"
                    y2="23"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
        <p
          class="root"
        />
      </div>
    </div>
  </div>
  <div
    class="newPassword"
  >
    <div
      class="root"
    >
      <span
        class="label"
      >
        New Password
      </span>
      <span
        class="root"
      >
        <span
          class="input"
        >
          <input
            class="input"
            name="newPassword"
            type="password"
            value=""
          />
        </span>
        <span
          class="before"
        />
        <span
          class="after"
        >
          <button
            type="button"
          >
            <span
              class="content"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17.94 17.94A10.07 10.07 0 0 1 12 20c-7 0-11-8-11-8a18.45 18.45 0 0 1 5.06-5.94M9.9 4.24A9.12 9.12 0 0 1 12 4c7 0 11 8 11 8a18.5 18.5 0 0 1-2.16 3.19m-6.72-1.07a3 3 0 1 1-4.24-4.24"
                />
                <line
                  x1="1"
                  x2="23"
                  y1="1"
                  y2="23"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
      <p
        class="root"
      />
    </div>
  </div>
</DocumentFragment>
`;
